xml: add support for other charsets

brenol opened the pull request:
This should fix #164 .

Ended up following this comment on mxj:
```
// ------------------- NewMapXml & NewMapXmlReader ... -------------------------
// If XmlCharsetReader != nil, it will be used to decode the XML, if required.
// Note: if CustomDecoder != nil, then XmlCharsetReader is ignored;
// set the CustomDecoder attribute instead.
//   import (
//	     charset "code.google.com/p/go-charset/charset"
//	     github.com/clbanning/mxj
//	 )
//   ...
//   mxj.XmlCharsetReader = charset.NewReader
//   m, merr := mxj.NewMapXml(xmlValue)
var XmlCharsetReader func(charset string, input io.Reader) (io.Reader, error)
```

And so I added a fork to be used inside dasel, as it seems to be pretty complete and a direct port of google code go-charset.

TomWright commented:
Hey @brenol, thanks for raising this PR - it looks good to me.

The only thing I'll ask for is some tests against different XML charsets to make sure things work as expected in a testable manner.

Could you add a couple of tests please?

brenol commented:
Hey @TomWright,

I've added a few test cases checking for the support of a few different encodings in XML.

Also, I changed from `go-charset` to  `x/net/charset`, as go-charset was panicking when reading UTF-16, and `x/net/charset` supports more charsets (and I expect a little more support from a x/ package).

But it seems that even with the correct encodings we might have problems with UTF16-BOM:
https://github.com/golang/go/issues/38335
(Basically, it seems encoding/xml reads the first line as UTF-8, and BOM makes it break).

I'm not sure if we should handle this here, though.

TomWright commented:
Looks good to me. I tested locally with the contents of https://github.com/TomWright/dasel/issues/164 and it worked as expected.

Thanks for the contribution 🙂
